timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampService.java

```diff
@@ -58,15 +60,29 @@ public static ClientAwareManagedTimestampService createDefault(ManagedTimestampS
 
     @Override
     public TimestampAndPartition getFreshTimestampForClient(UUID clientIdentifier) {
+        TimestampRangeAndPartition timestampRangeAndPartition = getFreshTimestampsForClient(clientIdentifier, 1);
+        return TimestampAndPartition.of(
+                timestampRangeAndPartition.range().getLowerBound(),
+                timestampRangeAndPartition.partition());
+    }
+
+    @Override
+    public TimestampRangeAndPartition getFreshTimestampsForClient(UUID clientIdentifier, int numTransactionsRequested) {
         while (true) {
-            TimestampRange timestampRange = delegate.getFreshTimestamps(NUM_PARTITIONS);
+            TimestampRange timestampRange = delegate.getFreshTimestamps(NUM_PARTITIONS * numTransactionsRequested);
```

Comment:
does this correctly handle the case where you asked for more than 10k timestamps and we didn't hand you 10k timestamps?

Reply:
I guess yes.

Reply:
I think so (and also the cases where you ask for `N < 10000` and get `M < N`, which IIRC can happen as leaders just return what they have rather than do more Paxos if they get to the end of their range) - the general idea is if you have something, _anything_ usable then return that, otherwise try again.

Reply:
I don't think your IIRC is quite right; looking at the Paxos bound it will increase the bound before handing you anything back.